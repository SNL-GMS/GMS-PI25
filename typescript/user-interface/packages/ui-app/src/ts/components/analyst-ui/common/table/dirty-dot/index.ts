export type { UnsavedChangesRow } from './dirty-dot-cell-renderer';
export type { DirtyDotCellRendererParams } from './dirty-dot-cell-renderer';
export { DirtyDotCellRenderer } from './dirty-dot-cell-renderer';
export { DirtyDotHeaderComponent } from './dirty-dot-header';
export { dirtyDotCellComparator } from './dirty-dot-util';
