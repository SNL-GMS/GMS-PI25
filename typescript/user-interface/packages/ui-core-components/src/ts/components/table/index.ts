export { ClientSideTable as Table } from './client-side-table';
export * from './infinite-table';
export * from './percent-bar';
export * from './table-cell-renderer';
export * from './types';
export * from './use-datasource';
export * as AgGridCommunity from 'ag-grid-community';
export { AgGridReact } from 'ag-grid-react';
