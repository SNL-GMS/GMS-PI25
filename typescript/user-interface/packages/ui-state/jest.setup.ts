/* eslint-disable max-classes-per-file */
/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable import/first */
/* eslint-disable import/no-extraneous-dependencies */
import 'fake-indexeddb/auto';

import type { FilterDefinition } from '@gms/common-model/lib/filter/types';
import crypto from 'crypto';
import * as util from 'util';

// This is to fix an issue with test leaks related to createStateSyncMiddleware() in jsdom
process.env.GMS_DISABLE_REDUX_STATE_SYNC = 'true';

Object.defineProperty(global, 'crypto', {
  value: {
    subtle: {
      digest: async (algorithm: AlgorithmIdentifier, msg: Uint8Array) =>
        (crypto.webcrypto as any).subtle.digest(algorithm, msg) as Promise<any>
    }
  }
});

Object.defineProperty(window, 'TextEncoder', {
  writable: true,
  value: util.TextEncoder
});
Object.defineProperty(window, 'TextDecoder', {
  writable: true,
  value: util.TextDecoder
});
Object.defineProperty(global, 'TextEncoder', {
  writable: true,
  value: util.TextEncoder
});
Object.defineProperty(global, 'TextDecoder', {
  writable: true,
  value: util.TextDecoder
});

import Adapter from '@cfaester/enzyme-adapter-react-18';
import Enzyme, { mount, render, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import type { GlobalWithFetchMock } from 'jest-fetch-mock';

const customGlobal: GlobalWithFetchMock = (global as unknown) as GlobalWithFetchMock;
// eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-require-imports
customGlobal.fetch = require('jest-fetch-mock');

customGlobal.fetchMock = customGlobal.fetch;

// eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-require-imports
require('jest-canvas-mock');

Enzyme.configure({ adapter: new Adapter() });

const globalAny: any = global;

// eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-require-imports
globalAny.fetch = require('jest-fetch-mock');

// Make Enzyme functions available in all test files without importing
globalAny.shallow = shallow;
globalAny.render = render;
globalAny.mount = mount;
globalAny.toJson = toJson;
globalAny.TextEncoder = util.TextEncoder;

globalAny.window = window;

class Worker {
  public url: string;

  public onmessage: (message: Record<string, unknown>, transfer?: Transferable[]) => void;

  public constructor(stringUrl: string) {
    this.url = stringUrl;
    this.onmessage = () => {
      // do nothing
    };
  }

  public postMessage(message: Record<string, unknown>, transfer?: Transferable[]) {
    this.onmessage(message, transfer);
  }
}
globalAny.window.Worker = Worker;

class MessagePort {
  public url: string;

  public onmessage: (message: Record<string, unknown>, transfer?: Transferable[]) => void;

  public postMessage(message: Record<string, unknown>, transfer?: Transferable[]) {
    this.onmessage(message, transfer);
  }
}

class SharedWorker {
  public url: string;

  public port = new MessagePort();

  public constructor(stringUrl: string) {
    this.url = stringUrl;

    this.port.onmessage = () => {
      // do nothing
    };
  }
}
globalAny.window.SharedWorker = SharedWorker;

// eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-require-imports,  import/no-extraneous-dependencies
require('jest-fetch-mock').enableMocks();

jest.mock('@gms/ui-wasm', () => {
  const isDesigned = jest.requireActual('@gms/ui-wasm/lib/gms-filters-provider/is-designed');

  const filter = async (
    filterDefinition: FilterDefinition,
    data: Float64Array
  ): Promise<Float64Array> => Promise.resolve(data);

  const design = async (filterDefinition: FilterDefinition): Promise<FilterDefinition> =>
    Promise.resolve(filterDefinition);

  const getBoundsForPositionBuffer = async (): Promise<{
    min: number;
    minSecs: number;
    max: number;
    maxSecs: number;
  }> => Promise.resolve({ min: 0, minSecs: 0, max: 10, maxSecs: 10 });

  return {
    design,
    filter,
    getBoundsForPositionBuffer,
    isDesigned: isDesigned.isDesigned
  };
});
