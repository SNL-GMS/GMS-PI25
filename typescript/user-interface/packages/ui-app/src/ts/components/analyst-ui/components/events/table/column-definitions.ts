import { EventsColumn } from '@gms/ui-state';

import { getConflictColumnDef } from '~analyst-ui/common/table/conflict-marker/conflict-marker-column-def';
import { getDirtyDotColumnDef } from '~analyst-ui/common/table/dirty-dot/dirty-dot-column-def';

import { eventColumnDisplayStrings } from '../types';
import { activeAnalystsColumnDef } from './columns/active-analysts';
import { confidenceSemiMajorColumnDef } from './columns/confidence-semi-major';
import { confidenceSemiMinorColumnDef } from './columns/confidence-semi-minor';
import { coverageSemiMajorColumnDef } from './columns/coverage-semi-major';
import { coverageSemiMinorColumnDef } from './columns/coverage-semi-minor';
import { deletedColumnDef } from './columns/deleted';
import { depthColumnDef } from './columns/depth';
import { depthUncertaintyColumnDef } from './columns/depth-uncertainty';
import { latitudeColumnDef } from './columns/latitude';
import { longitudeColumnDef } from './columns/longitude';
import { mbColumnDef } from './columns/mb';
import { mlColumnDef } from './columns/ml';
import { msColumnDef } from './columns/ms';
import { numberAssociatedColumnDef } from './columns/number-associated';
import { numberDefiningColumnDef } from './columns/number-defining';
import { observationsStandardDeviationColumnDef } from './columns/observerations-standard-dev';
import { preferredColumnDef } from './columns/preferred';
import { regionColumnDef } from './columns/region';
import { rejectedColumnDef } from './columns/rejected';
import { statusColumnDef } from './columns/status';
import { timeColumnDef } from './columns/time';
import { timeUncertaintyColumnDef } from './columns/time-uncertainty';

/**
 * @returns List of column definition objects for the Events Table
 */
export const getEventsTableColumnDefs = () => {
  return [
    getDirtyDotColumnDef(
      eventColumnDisplayStrings.get(EventsColumn.unsavedChanges),
      EventsColumn.unsavedChanges,
      eventColumnDisplayStrings.get(EventsColumn.unsavedChanges)
    ),
    getConflictColumnDef(
      eventColumnDisplayStrings.get(EventsColumn.conflict),
      EventsColumn.conflict,
      eventColumnDisplayStrings.get(EventsColumn.conflict)
    ),
    timeColumnDef,
    timeUncertaintyColumnDef,
    latitudeColumnDef,
    longitudeColumnDef,
    depthColumnDef,
    depthUncertaintyColumnDef,
    mbColumnDef,
    msColumnDef,
    mlColumnDef,
    numberAssociatedColumnDef,
    numberDefiningColumnDef,
    observationsStandardDeviationColumnDef,
    coverageSemiMajorColumnDef,
    coverageSemiMinorColumnDef,
    confidenceSemiMajorColumnDef,
    confidenceSemiMinorColumnDef,
    regionColumnDef,
    activeAnalystsColumnDef,
    preferredColumnDef,
    statusColumnDef,
    rejectedColumnDef,
    deletedColumnDef
  ];
};
