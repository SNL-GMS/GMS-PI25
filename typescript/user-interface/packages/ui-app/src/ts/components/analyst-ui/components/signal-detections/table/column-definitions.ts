import { SignalDetectionColumn } from '@gms/ui-state';

import { getConflictColumnDef } from '~analyst-ui/common/table/conflict-marker/conflict-marker-column-def';
import { getDirtyDotColumnDef } from '~analyst-ui/common/table/dirty-dot/dirty-dot-column-def';

import { signalDetectionColumnDisplayStrings } from '../types';
import { amplitudeColumnDef } from './columns/amplitude';
import { assocStatusColumnDef } from './columns/assoc-status';
import { azimuthColumnDef } from './columns/azimuth';
import { azimuthStdDevColumnDef } from './columns/azimuth-std-dev';
import { channelColumnDef } from './columns/channel';
import { deletedColumnDef } from './columns/deleted';
import { emergenceAngleColumnDef } from './columns/emergence-angle';
import { longPeriodFirstMotionColumnDef } from './columns/long-period-first-motion';
import { periodColumnDef } from './columns/period';
import { phaseColumnDef } from './columns/phase';
import { phaseConfidenceColumnDef } from './columns/phase-confidence';
import { rectilinearityColumnDef } from './columns/rectilinearity';
import { shortPeriodFirstMotionColumnDef } from './columns/short-period-first-motion';
import { slownessColumnDef } from './columns/slowness';
import { slownessStandardDeviationColumnDef } from './columns/slowness-std-dev';
import { sNRColumnDef } from './columns/snr';
import { stationColumnDef } from './columns/station';
import { timeColumnDef } from './columns/time';
import { timeStandardDeviationColumnDef } from './columns/time-std-dev';

/**
 * @returns List of column definition objects for the Signal Detection Table
 */
export const getSignalDetectionTableColumnDefs = () => {
  return [
    getDirtyDotColumnDef(
      signalDetectionColumnDisplayStrings.get(SignalDetectionColumn.unsavedChanges),
      SignalDetectionColumn.unsavedChanges,
      signalDetectionColumnDisplayStrings.get(SignalDetectionColumn.unsavedChanges)
    ),
    assocStatusColumnDef,
    getConflictColumnDef(
      signalDetectionColumnDisplayStrings.get(SignalDetectionColumn.conflict),
      SignalDetectionColumn.conflict,
      signalDetectionColumnDisplayStrings.get(SignalDetectionColumn.conflict)
    ),
    stationColumnDef,
    channelColumnDef,
    phaseColumnDef,
    phaseConfidenceColumnDef,
    timeColumnDef,
    timeStandardDeviationColumnDef,
    azimuthColumnDef,
    azimuthStdDevColumnDef,
    slownessColumnDef,
    slownessStandardDeviationColumnDef,
    amplitudeColumnDef,
    periodColumnDef,
    sNRColumnDef,
    rectilinearityColumnDef,
    emergenceAngleColumnDef,
    shortPeriodFirstMotionColumnDef,
    longPeriodFirstMotionColumnDef,
    deletedColumnDef
  ];
};
