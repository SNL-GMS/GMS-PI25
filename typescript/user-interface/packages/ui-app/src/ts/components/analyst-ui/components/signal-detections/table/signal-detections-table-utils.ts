import type { CommonTypes } from '@gms/common-model';
import { SignalDetectionTypes } from '@gms/common-model';
import {
  findAmplitudeFeatureMeasurementValue,
  findAzimuthFeatureMeasurementValue,
  findEmergenceAngleFeatureMeasurementValue,
  findFeatureMeasurementChannelName,
  findLongPeriodFirstMotionFeatureMeasurementValue,
  findRectilinearityFeatureMeasurementValue,
  findShortPeriodFirstMotionFeatureMeasurementValue,
  findSlownessFeatureMeasurementValue
} from '@gms/common-model/lib/signal-detection/util';
import { formatTimeForDisplay } from '@gms/common-util';
import type { AgGridCommunity, RowNode } from '@gms/ui-core-components';
import type {
  DisplayedSignalDetectionConfigurationEnum,
  SignalDetectionFetchResult
} from '@gms/ui-state';
import { SignalDetectionColumn } from '@gms/ui-state';
import type { AgGridReact } from 'ag-grid-react';
import Immutable from 'immutable';
import includes from 'lodash/includes';

import {
  getEdgeType,
  shouldDisplaySignalDetection
} from '~analyst-ui/common/utils/signal-detection-util';
import { EventFilterOptions } from '~analyst-ui/components/events/types';
import type { SignalDetectionRow } from '~analyst-ui/components/signal-detections/types';
import {
  formatNumberForDisplayFixedThreeDecimalPlaces,
  getTableCellStringValue,
  setRowNodeSelection
} from '~common-ui/common/table-utils';

import {
  nonIdealStateLoadingSignalDetections,
  nonIdealStateNoSignalDetections,
  nonIdealStateNoSignalDetectionsErrorState,
  nonIdealStateNoSignalDetectionsSyncedTimeRange
} from '../signal-detections-non-ideal-states';

/**
 * Determines which non-ideal state, if any, should be presented.
 *
 * @returns NonIdealState, or undefined if not needed.
 */
export const handleNonIdealState = (
  signalDetectionsQuery: SignalDetectionFetchResult,
  isSynced: boolean
): JSX.Element | undefined => {
  if (signalDetectionsQuery.isLoading) {
    return nonIdealStateLoadingSignalDetections;
  }

  if (
    !signalDetectionsQuery.isLoading &&
    !signalDetectionsQuery.isError &&
    signalDetectionsQuery.data?.length === 0
  )
    return isSynced
      ? nonIdealStateNoSignalDetectionsSyncedTimeRange
      : nonIdealStateNoSignalDetections;

  if (signalDetectionsQuery.isError && signalDetectionsQuery.data?.length === 0)
    return nonIdealStateNoSignalDetectionsErrorState;

  return undefined;
};

/**
 * Type guard for {@link SignalDetectionRow}
 */
export function isSignalDetectionRow(object: unknown): object is SignalDetectionRow {
  return (object as SignalDetectionRow).assocStatus !== undefined;
}

/**
 * In charge of setting the row class
 */
export const sdRowClassRules: AgGridCommunity.RowClassRules = {
  'edge-SD-row': params => params.data.edgeType !== EventFilterOptions.INTERVAL,
  'action-target-row': (params: { data: SignalDetectionRow }) => params.data.isActionTarget,
  'unqualified-action-target-row': (params: { data: SignalDetectionRow }) =>
    params.data.isUnqualifiedActionTarget
};

/**
 * This is the set of default columns to be displayed in the SD table
 * This object gets updated when columns are selected/deselected in the column picker so that state
 * doesn't get lost
 */
export const signalDetectionsColumnsToDisplay: Immutable.Map<
  SignalDetectionColumn,
  boolean
> = Immutable.Map([
  ...Object.values(SignalDetectionColumn).map<[SignalDetectionColumn, boolean]>(v => [v, true]),
  [SignalDetectionColumn.phaseConfidence, false],
  [SignalDetectionColumn.rectilinearity, false],
  [SignalDetectionColumn.emergenceAngle, false],
  [SignalDetectionColumn.shortPeriodFirstMotion, false],
  [SignalDetectionColumn.longPeriodFirstMotion, false]
]);

/**
 * These two fields have in-band error codes, if we see a '-1', we need to instead display "Unknown"
 * these fields are otherwise treated as the rest of the numeric fields
 *
 * @param value
 */
export function formatRectilinearityOrEmergenceForDisplay(value: number): string {
  if (value === -1) return 'Unknown';
  return formatNumberForDisplayFixedThreeDecimalPlaces(value);
}

/**
 * Called by ag-grid to determine if an external filter is present/active.
 */
export function agGridIsExternalFilterPresent(
  signalDetectionFilterState: Record<DisplayedSignalDetectionConfigurationEnum, boolean>
): boolean {
  const {
    signalDetectionUnassociated,
    signalDetectionAssociatedToOpenEvent,
    signalDetectionAssociatedToCompletedEvent,
    signalDetectionAssociatedToOtherEvent,
    signalDetectionDeleted,
    signalDetectionConflicts,
    signalDetectionBeforeInterval,
    signalDetectionAfterInterval
  } = signalDetectionFilterState;

  // If any of the event associations are unchecked then the table must be filtered.
  return (
    !signalDetectionUnassociated ||
    !signalDetectionAssociatedToOpenEvent ||
    !signalDetectionAssociatedToCompletedEvent ||
    !signalDetectionAssociatedToOtherEvent ||
    !signalDetectionConflicts ||
    !signalDetectionDeleted ||
    !signalDetectionBeforeInterval ||
    !signalDetectionAfterInterval
  );
}

/**
 * Passed to ag-grid. Should return true if external filter passes, otherwise false.
 */
export function agGridDoesExternalFilterPass(
  node: RowNode,
  signalDetectionFilterState: Record<DisplayedSignalDetectionConfigurationEnum, boolean>
): boolean {
  return shouldDisplaySignalDetection(
    node.data.assocStatus,
    node.data.edgeType,
    node.data.conflict,
    signalDetectionFilterState
  );
}

/**
 * Builds a single {@link SignalDetectionRow} given a {@link SignalDetectionTypes.SignalDetection} object
 *
 * @param sdDataForRow Object containing like data related to the signal detection.
 * @param timeRange Used to determine if SD is an edge SD or not
 */
export function buildSignalDetectionRow(
  sdDataForRow: {
    sd: SignalDetectionTypes.SignalDetection;
    associationStatus: SignalDetectionTypes.SignalDetectionStatus;
    sdInConflict: boolean;
    sdIsActionTarget: boolean;
    validActionTargetSignalDetectionIds: string[];
  },
  timeRange: CommonTypes.TimeRange
): SignalDetectionRow {
  const {
    sd,
    associationStatus,
    sdInConflict,
    sdIsActionTarget,
    validActionTargetSignalDetectionIds
  } = sdDataForRow;
  const featureMeasurements = SignalDetectionTypes.Util.getCurrentHypothesis(
    sd.signalDetectionHypotheses
  )?.featureMeasurements;

  return {
    id: sd.id,
    unsavedChanges: !!(sd._uiHasUnsavedEventSdhAssociation || sd._uiHasUnsavedChanges),
    assocStatus: associationStatus,
    conflict: sdInConflict,
    isActionTarget: sdIsActionTarget,
    isUnqualifiedActionTarget:
      sdIsActionTarget && !includes(validActionTargetSignalDetectionIds, sd.id),
    station: getTableCellStringValue(sd?.station.name),
    channel: findFeatureMeasurementChannelName(featureMeasurements),
    phase: getTableCellStringValue(
      SignalDetectionTypes.Util.findPhaseFeatureMeasurementValue(featureMeasurements)?.value
    ),
    phaseConfidence: formatNumberForDisplayFixedThreeDecimalPlaces(
      SignalDetectionTypes.Util.findPhaseFeatureMeasurementValue(featureMeasurements)?.confidence
    ),
    time: formatTimeForDisplay(
      SignalDetectionTypes.Util.findArrivalTimeFeatureMeasurementValue(featureMeasurements)
        ?.arrivalTime?.value
    ),
    timeStandardDeviation: formatNumberForDisplayFixedThreeDecimalPlaces(
      SignalDetectionTypes.Util.findArrivalTimeFeatureMeasurementValue(featureMeasurements)
        ?.arrivalTime?.standardDeviation
    ).replace(/,/g, ''),
    azimuth: formatNumberForDisplayFixedThreeDecimalPlaces(
      findAzimuthFeatureMeasurementValue(featureMeasurements)?.measuredValue?.value
    ),
    azimuthStandardDeviation: formatNumberForDisplayFixedThreeDecimalPlaces(
      findAzimuthFeatureMeasurementValue(featureMeasurements)?.measuredValue?.standardDeviation
    ),
    slowness: formatNumberForDisplayFixedThreeDecimalPlaces(
      findSlownessFeatureMeasurementValue(featureMeasurements)?.measuredValue?.value
    ),
    slownessStandardDeviation: formatNumberForDisplayFixedThreeDecimalPlaces(
      findSlownessFeatureMeasurementValue(featureMeasurements)?.measuredValue?.standardDeviation
    ),
    amplitude: formatNumberForDisplayFixedThreeDecimalPlaces(
      findAmplitudeFeatureMeasurementValue(
        featureMeasurements,
        SignalDetectionTypes.FeatureMeasurementType.AMPLITUDE_A5_OVER_2
      )?.amplitude?.value
    ),
    period: formatNumberForDisplayFixedThreeDecimalPlaces(
      findAmplitudeFeatureMeasurementValue(
        featureMeasurements,
        SignalDetectionTypes.FeatureMeasurementType.AMPLITUDE_A5_OVER_2
      )?.period
    ),
    sNR: formatNumberForDisplayFixedThreeDecimalPlaces(
      SignalDetectionTypes.Util.findArrivalTimeFeatureMeasurement(featureMeasurements)?.snr?.value
    ),
    rectilinearity: formatRectilinearityOrEmergenceForDisplay(
      findRectilinearityFeatureMeasurementValue(featureMeasurements)?.measuredValue?.value
    ),
    emergenceAngle: formatRectilinearityOrEmergenceForDisplay(
      findEmergenceAngleFeatureMeasurementValue(featureMeasurements)?.measuredValue?.value
    ),
    shortPeriodFirstMotion: getTableCellStringValue(
      findShortPeriodFirstMotionFeatureMeasurementValue(featureMeasurements)?.value
    ),
    longPeriodFirstMotion: getTableCellStringValue(
      findLongPeriodFirstMotionFeatureMeasurementValue(featureMeasurements)?.value
    ),
    deleted: SignalDetectionTypes.Util.getCurrentHypothesis(sd?.signalDetectionHypotheses).deleted
      ? 'True'
      : 'False',
    edgeType: getEdgeType(
      timeRange,
      SignalDetectionTypes.Util.findArrivalTimeFeatureMeasurementValue(featureMeasurements)
        ?.arrivalTime?.value
    )
  };
}

/**
 * Cycles through all rows in the table and updates their row selection attribute
 * Sets rowSelection to true if their row id is in {@param selectedSdIds}, false otherwise
 *
 * @param tableRef
 * @param selectedSdIds
 */
export function updateRowSelection(
  tableRef: React.MutableRefObject<AgGridReact>,
  selectedSdIds: string[]
): React.MutableRefObject<AgGridReact> {
  tableRef.current.api.forEachNode(node =>
    setRowNodeSelection(node, selectedSdIds.includes(node.id))
  );
  return tableRef;
}

/**
 * Checks for 0 to 4 integer digits before optional period and 0 to 8 fraction digits
 * Trailing period without fraction digits is also permitted
 *
 * @param stdDev
 * @returns boolean from regex test
 */
export const isValidSdListStdDev = (stdDev: string) => {
  const regex = /^\d{0,4}(?:\.\d{0,8})?\.?$/g;

  return regex.test(stdDev);
};
