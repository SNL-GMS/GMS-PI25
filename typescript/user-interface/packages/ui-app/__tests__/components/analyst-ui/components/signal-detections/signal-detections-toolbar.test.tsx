/* eslint-disable react/jsx-no-constructed-context-values */
import { getStore } from '@gms/ui-state';
import { render } from '@testing-library/react';
import * as React from 'react';
import { Provider } from 'react-redux';

import { signalDetectionsColumnsToDisplay } from '../../../../../src/ts/components/analyst-ui/components/signal-detections/table/signal-detections-table-utils';
import {
  showDetectionsOnChange,
  SignalDetectionsToolbar
} from '../../../../../src/ts/components/analyst-ui/components/signal-detections/toolbar/signal-detections-toolbar';
import { BaseDisplayContext } from '../../../../../src/ts/components/common-ui/components/base-display/base-display-context';

const mockSetSelectedSDColumnsToDisplay = jest.fn();
const mockSetCreateEventMenuVisibility = jest.fn();

describe('Signal Detection Toolbar', () => {
  test('can be mounted', () => {
    expect(SignalDetectionsToolbar).toBeDefined();
  });

  test('should match snapshot', () => {
    const { container } = render(
      <Provider store={getStore()}>
        <BaseDisplayContext.Provider
          value={{
            glContainer: {} as any,
            widthPx: 100,
            heightPx: 100
          }}
        >
          <SignalDetectionsToolbar
            countEntryRecord={{
              Total: { count: 0, color: 'tomato', isShown: true, tooltip: 'test' },
              Completed: { count: 0, color: 'tomato', isShown: true, tooltip: 'test' },
              Deleted: { count: 0, color: 'tomato', isShown: true, tooltip: 'test' },
              Open: { count: 0, color: 'tomato', isShown: true, tooltip: 'test' },
              Other: { count: 0, color: 'tomato', isShown: true, tooltip: 'test' },
              Conflicts: { count: 0, color: 'tomato', isShown: true, tooltip: 'test' },
              Unassociated: { count: 0, color: 'tomato', isShown: true, tooltip: 'test' }
            }}
            setSelectedSDColumnsToDisplay={mockSetSelectedSDColumnsToDisplay}
            selectedSDColumnsToDisplay={signalDetectionsColumnsToDisplay}
            setCreateEventMenuVisibility={mockSetCreateEventMenuVisibility}
          />
        </BaseDisplayContext.Provider>
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });

  test('checkbox list showDetectionsOnChange with configuration arguments', () => {
    const dispatch = jest.fn();
    const eventObject = {
      syncWaveform: false,
      signalDetectionBeforeInterval: true,
      signalDetectionAfterInterval: true,
      signalDetectionAssociatedToOpenEvent: true,
      signalDetectionAssociatedToCompletedEvent: true,
      signalDetectionAssociatedToOtherEvent: true,
      signalDetectionConflicts: true,
      signalDetectionUnassociated: false
    };
    showDetectionsOnChange(dispatch)(eventObject);
    expect(dispatch).toHaveBeenCalledTimes(1);
  });
});
