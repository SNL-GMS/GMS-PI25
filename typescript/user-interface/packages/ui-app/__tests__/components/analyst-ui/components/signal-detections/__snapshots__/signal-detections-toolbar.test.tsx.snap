// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signal Detection Toolbar should match snapshot 1`] = `
<div>
  <div
    class="toolbar"
    style="width: 84px;"
  >
    <div
      class="toolbar__left-group"
    >
      <div
        class="toolbar-item toolbar-item__left"
      >
        <div
          class="counter"
        >
          <span
            class="counter__entry"
            style="--entry-color: tomato;"
            title="test"
          >
            <span
              class="counter__label"
            >
              Total
            </span>
            <span
              class="counter__count "
            >
              <span
                class="counter__count--prefix"
              >
                0
                0
              </span>
              <span
                class="counter__count--main"
              >
                0
              </span>
            </span>
          </span>
          <span
            class="counter__entry"
            style="--entry-color: tomato;"
            title="test"
          >
            <span
              class="counter__label"
            >
              Completed
            </span>
            <span
              class="counter__count "
            >
              <span
                class="counter__count--prefix"
              >
                0
                0
              </span>
              <span
                class="counter__count--main"
              >
                0
              </span>
            </span>
          </span>
          <span
            class="counter__entry"
            style="--entry-color: tomato;"
            title="test"
          >
            <span
              class="counter__label"
            >
              Deleted
            </span>
            <span
              class="counter__count "
            >
              <span
                class="counter__count--prefix"
              >
                0
                0
              </span>
              <span
                class="counter__count--main"
              >
                0
              </span>
            </span>
          </span>
          <span
            class="counter__entry"
            style="--entry-color: tomato;"
            title="test"
          >
            <span
              class="counter__label"
            >
              Open
            </span>
            <span
              class="counter__count "
            >
              <span
                class="counter__count--prefix"
              >
                0
                0
              </span>
              <span
                class="counter__count--main"
              >
                0
              </span>
            </span>
          </span>
          <span
            class="counter__entry"
            style="--entry-color: tomato;"
            title="test"
          >
            <span
              class="counter__label"
            >
              Other
            </span>
            <span
              class="counter__count "
            >
              <span
                class="counter__count--prefix"
              >
                0
                0
              </span>
              <span
                class="counter__count--main"
              >
                0
              </span>
            </span>
          </span>
          <span
            class="counter__entry"
            style="--entry-color: tomato;"
            title="test"
          >
            <span
              class="counter__label"
            >
              Conflicts
            </span>
            <span
              class="counter__count "
            >
              <span
                class="counter__count--prefix"
              >
                0
                0
              </span>
              <span
                class="counter__count--main"
              >
                0
              </span>
            </span>
          </span>
          <span
            class="counter__entry"
            style="--entry-color: tomato;"
            title="test"
          >
            <span
              class="counter__label"
            >
              Unassociated
            </span>
            <span
              class="counter__count "
            >
              <span
                class="counter__count--prefix"
              >
                0
                0
              </span>
              <span
                class="counter__count--main"
              >
                0
              </span>
            </span>
          </span>
        </div>
      </div>
    </div>
    <div
      class="toolbar__center-group"
    >
      <div
        class="toolbar__whitespace"
        style="width: 84px;"
      />
    </div>
    <div
      class="toolbar__right-group"
    >
      <div
        class="toolbar-item
                  "
      >
        <div>
          <div
            class="bp4-context-menu2"
          >
            <div />
          </div>
          <div
            data-cy="sd-table-detections-toggle"
          >
            <button
              class="bp4-button bp4-align-left toolbar-button"
              title="Set which detections are visible"
              type="button"
            >
              <span
                class="bp4-button-text"
              >
                <span>
                  Show Detections
                </span>
                <span
                  aria-hidden="true"
                  class="bp4-icon bp4-icon-chevron-down"
                  icon="chevron-down"
                >
                  <svg
                    data-icon="chevron-down"
                    height="16"
                    role="img"
                    viewBox="0 0 16 16"
                    width="16"
                  >
                    <path
                      d="M12 5c-.28 0-.53.11-.71.29L8 8.59l-3.29-3.3a1.003 1.003 0 00-1.42 1.42l4 4c.18.18.43.29.71.29s.53-.11.71-.29l4-4A1.003 1.003 0 0012 5z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
              </span>
            </button>
          </div>
        </div>
      </div>
      <div
        class="toolbar-item
                  "
      >
        <div>
          <div
            class="bp4-context-menu2"
          >
            <div />
          </div>
          <div
            data-cy="sd-table-column-picker"
          >
            <button
              class="bp4-button bp4-align-left toolbar-button"
              title="Set which columns are visible"
              type="button"
            >
              <span
                class="bp4-button-text"
              >
                <span>
                  Show Columns
                </span>
                <span
                  aria-hidden="true"
                  class="bp4-icon bp4-icon-chevron-down"
                  icon="chevron-down"
                >
                  <svg
                    data-icon="chevron-down"
                    height="16"
                    role="img"
                    viewBox="0 0 16 16"
                    width="16"
                  >
                    <path
                      d="M12 5c-.28 0-.53.11-.71.29L8 8.59l-3.29-3.3a1.003 1.003 0 00-1.42 1.42l4 4c.18.18.43.29.71.29s.53-.11.71-.29l4-4A1.003 1.003 0 0012 5z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
              </span>
            </button>
          </div>
        </div>
      </div>
      <div
        class="toolbar-item
                  "
      >
        <div>
          <button
            class="bp4-button bp4-align-left toolbar-button"
            data-cy="btn-create-event"
            title="Create new event from selected signal detections or virtual event when no signal detections selected (hot key: undefined)"
            type="button"
          >
            <span
              class="bp4-button-text"
            >
              <span>
                Create Event
              </span>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
